### PR TITLE
Trie

### PR DESCRIPTION
**Pull request by SabrinaSnider**
Designed a Trie data structure to be used for autocomplete. A Trie is a tree-based data structure used for storing word prefixes and for quick retrievals.

The root of the Trie stores an empty string. Each of the root's children (up to 26 children- one for each letter) will store a 1-letter string. Each of those node's children (again, up to 26 children) will store a 2-letter string. The further down the Trie you go, the longer the strings at each node.

Each node also has a boolean to mark whether that string is a completed word, and should therefore be considered for an autocomplete suggestion.

I'm requesting reviews on this PR to make the merge of the autocomplete feature to master later be easier.

### CODE REVIEW

**Comment by brucezheng** [comment]
🏆🏆🏆 Achievement unlocked: Algorithmic complexity 🏆🏆🏆

File: public/js/discover/trie.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
+/* eslint-disable no-unused-vars */
+
+const categories = [
```
Comment (danielabreo) [inline public/js/discover/trie.js:3]:
```suggestion
const categories = [
```

Comment (danielabreo) [reply public/js/discover/trie.js:3]: nit: delete first two lines and use suggested fix for the `no-unused-vars` error.

File: public/js/discover/trie.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+]
+
+const categoryTrie = new Trie(categories);
+
+/**
+ * Gives autocomplete suggestions for item categories.
+ * @param {string} input The prefix to base autocomplete suggestions on.
+ * @return {array} The list of autocomplete suggestions.
+ */
+function getAutocompleteCategories(input) {
```
Comment (danielabreo) [inline public/js/discover/trie.js:29]:
```suggestion
window.getAutocompleteCategories = function getAutocompleteCategories(input) {
```

Comment (danielabreo) [reply public/js/discover/trie.js:29]: this will solve the `no-unused-variables` error and should keep the same functionality.

Comment (SabrinaSnider) [reply public/js/discover/trie.js:29]: Should I do this even if I don't use `getAutocompleteCategories()` in the HTML file? I was planning on calling it from another js file.

Comment (danielabreo) [reply public/js/discover/trie.js:29]: yes I would still recommend doing so.

File: public/js/discover/trie.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  constructor(words) {
+    this.root = new TrieNode('');
+    for (const word of words) this.insert(word);
+  }
+
+  /**
+   * Adds the word to the Trie, creating new nodes for its prefixes if needed.
+   * @param {string} word The word to insert into the tree.
+   */
+  insert(word) {
```
Comment (brucezheng) [inline public/js/discover/trie.js:63]: Do we remove special characters for this?

File: public/js/discover/trie.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+      if (currentNode.children[index]) {
+        currentNode = currentNode.children[index];
+      } else {
+        // If the prefix isn't in the tree, return no suggestions.
+        return [];
+      }
+    }
+    return this.getChildWords(currentNode);
+  }
```
Comment (brucezheng) [inline public/js/discover/trie.js:102]: Trie is a data structure, so I think ```autocomplete``` should not be a class method if you just want a generic data structure implementation. Instead, we could implement this function as ```find``` which returns the ```TrieNode``` and then can just use ```getChildren``` on that return value.

File: public/js/discover/trie.js (hunk 3 above)
Comment (SabrinaSnider) [reply public/js/discover/trie.js:63]: Added. 👍

File: public/js/discover/trie.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  'books',
+  'school supplies',
+  "children's toys",
+  'hair products',
+  'deoderant',
+  'soap',
+  'bottled drinks',
+  'computers',
+  'cars',
+];
```
Comment (ccyang314) [inline public/js/discover/trie.js:18]: Question on this hardcoded list. Are we planning to switch to read this list from Firestore in post-MVP/MVP?

Comment (SabrinaSnider) [reply public/js/discover/trie.js:18]: Yep, we're aiming to read categories from the database for MVP 👍

File: public/js/discover/trie.js [hunk 6]
```diff
@@ -61,7 +61,7 @@ class Trie {
    * @param {string} word The word to insert into the tree.
    */
   insert(word) {
-    word = word.toLowerCase();
+    word = word.toLowerCase().replace(/[^a-zA-Z ]/g, '');
```
Comment (brucezheng) [inline public/js/discover/trie.js:64]: Move this to a separate method since you're sanitizing in multiple places.

**Pull request merged**